lib/Transforms/StagedPatternRewriteDriver.cpp

```diff
@@ -0,0 +1,289 @@
+//===- StagedPatternRewriteDriver.cpp - A staged rewriter -----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements mlir::applyStagedPatternsAndCanonicalize.
+//
+//===----------------------------------------------------------------------===//
+
+#include "Transforms/StagedPatternRewriteDriver.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Rewrite/PatternApplicator.h"
+#include "mlir/Transforms/FoldUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/ScopedPrinter.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace mlir;
+
+#define DEBUG_TYPE "staged-rewrite"
+
+namespace {
+
+void addToWorklist(Operation *op, std::vector<Operation *> &worklist,
+                   DenseMap<Operation *, unsigned> &worklistMap) {
+  // Check to see if the worklist already contains this op.
+  if (worklistMap.count(op))
+    return;
+
+  worklistMap[op] = worklist.size();
+  worklist.push_back(op);
+}
+
+Operation *popFromWorklist(std::vector<Operation *> &worklist,
+                           DenseMap<Operation *, unsigned> &worklistMap) {
+  auto *op = worklist.back();
+  worklist.pop_back();
+
+  // This operation is no longer in the worklist, keep worklistMap up to date.
+  if (op)
+    worklistMap.erase(op);
+  return op;
+}
+
+void removeFromWorklist(Operation *op, std::vector<Operation *> &worklist,
+                        DenseMap<Operation *, unsigned> &worklistMap) {
+  auto it = worklistMap.find(op);
+  if (it != worklistMap.end()) {
+    assert(worklist[it->second] == op && "malformed worklist data structure");
+    worklist[it->second] = nullptr;
+    worklistMap.erase(it);
+  }
+}
+
+const char *logLineComment =
+    "//++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++//\n";
+
+class SinglePatternListener : OpBuilder::Listener {
```

Comment:
This isn't set on the builder, and duplicates the worklists from the `PatternRewriter`. Can we just have the rewriter also derive OpBuilder::Listener instead? + set the listener.